rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+
+### Problems addressed
+
+#### Spaghetti dependencies
+
+Everything being in a single repository encourages everyone to freely use all other code in the repository, even when this does not make sense. This ensures the library is very difficult to modify later.
+
+#### Long build times
+
+Many volunteers and developers outside Google use their laptops for development. On such systems, TF development cycles require building all of tensorflow that takes around 2 hours. While there is an argument to be made that bazel caching should ensure that they build all of TF only once, without forge, bazel caching is not working as [well as expected](https://docs.bazel.build/versions/master/remote-caching.html#known-issues).
+
+#### Adding support for new hardware is very difficult and not scalable
+
+The ML ecosystem is constantly expanding. New hardware for accelerating ML applications is being worked on by many teams inside and outside Google. As the most popular machine learning framework, TF is expected to add support for many of these hardware as quickly as possible.
+
+Currently, this means that all such hardware developers need to check in their code into the [main tensorflow repository](http://github.com/tensorflow/tensorflow). This means that all the changes are required to go through TF team's review. This can make merging support for new hardware very very difficult.
+
+#### Long PR review queue
+
+TensorFlow is a very successful opensource project. It is the [4th most forked](https://github.com/search?o=desc&q=stars:%3E1&s=forks&type=Repositories) and [5th most starred](https://github.com/search?q=stars%3A%3E0&s=stars&type=Repositories) project on github. This also means that TF receives quite a lot of opensource contributions. The TensorFlow team has to review all contributions to TensorFlow itself. This creates a bottleneck for merging the changes to the main repository.
+
+#### Flexibility for collaborators
+
+Currently, any partner or contributor that would like to work with us are subject to all the rules within of the main repository. Some of these can be relaxed through modularization, where work can happen in a separate repository.
+
+#### Large TF support matrix
+
+The TF support story is a unique beast. Our support matrix has a lot of orthogonal dimensions. Below are some of the more prominent dimensions:
+
+*   Environment (google3, opensource)
+*   Operating system (Linux, windows, macos, mobile)
+*   Architecture (x86, arm, ppc64, …)
+*   Accelerator (CPU, GPU,  TPU)
+*   Compiler (GCC, Clang, MSVC)
+*   Python version (2, 3.4, 3.5, 3.6,  3.7)
+
+More can be added to this list where we have to rebuild TensorFlow to support different network architectures, CUDA versions, SIMD instruction sets, etc.
```

Comment:
How do we handle MKL PR by outsiders? What should Intel do?

Reply:
That would be entirely intel's call.
Intel may choose to completely close the source of the MKL plugin, or manage the plugin code just like another project, just like this one: https://github.com/intel/mkl-dnn